### PR TITLE
Random input feature dropping curve, model selection visualization [issue #1024]

### PR DESCRIPTION
**Pull request by charlesincharge**
## Summary

This PR addresses https://github.com/DistrictDataLabs/yellowbrick/issues/1204, which requested a feature-dropping-curve (also known as a neuron dropping curve in neural decoding).

I have made the following changes:

1. Added a module to `yellowbrick.model_selection` called `dropping_curve`. It is largely based on `yellowbrick.model_selection.validation_curve`

### Sample Code and Plot

```python
from sklearn.naive_bayes import MultinomialNB
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

from yellowbrick.datasets import load_game
from yellowbrick.model_selection.dropping_curve import dropping_curve

# Load Connect-4 game data
X, y = load_game()
X_enc = OneHotEncoder().fit_transform(X)
le = LabelEncoder()
y_enc = le.fit_transform(y)

dropping_curve(
    MultinomialNB(),
    X_enc,
    y_enc,
    feature_sizes=np.linspace(0.05, 1, 20),
)
```

Figure_1

### TODOs and questions

Still to do:

- Unit tests

Questions for the @DistrictDataLabs/team-oz-maintainers:

- This code generates ~30 additional blank figures when I run it through a Python script. Is that expected?

### CHECKLIST

- [ ] _Is the commit message formatted correctly?_
- [ ] _Have you noted the new functionality/bugfix in the release notes of the next release?_

- [x] _Included a sample plot to visually illustrate your changes?_
- [x] _Do all of your functions and methods have docstrings?_
- [x] _Have you added/updated unit tests where appropriate?_
- [x] _Have you updated the baseline images if necessary?_
- [x] _Have you run the unit tests using `pytest`?_
- [x] _Is your code style correct (are you using PEP8, pyflakes)?_
- [ ] _Have you documented your new feature/functionality in the docs?_

- [ ] _Have you built the docs using `make html`?_

### CODE REVIEW

**Comment by bbengfort** [comment]
@charlesincharge Happy New Year! And thank you for opening this PR into Yellowbrick! @lwgray @pdamodaran @Kautumn06 would any of you be interested in doing the code review for this?

**Comment by pdamodaran** [comment]
@bbengfort - apologies for the delay in responding.  I can review this PR next weekend.

**Comment by bbengfort** [comment]
Thanks @pdamodaran!

**Comment by lwgray** [comment]
@charlesincharge what version of matplotlib did you use when developing this feature?

**Comment by charlesincharge** [comment]
@lwgray I used `matplotlib          3.5.0`

**Comment by lwgray** [comment]
@charlesincharge Thanks for contributing such a wonderful visualizer.  I have spent quite a bit of time working on it because of conflicts with our recent upgrade to Sci-kit 1.0.  To help us get this up and merged, would you mind doing the documentation?  You can read more about it here https://www.scikit-yb.org/en/latest/contributing/developing_visualizers.html#documentation

Here is a recent PR that has docs added
https://github.com/DistrictDataLabs/yellowbrick/pull/1189

File: yellowbrick/model_selection/dropping_curve.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+            n_features is the number of features.
+        y : array-like, shape (n_samples) or (n_samples, n_features), optional
+            Target relative to X for classification or regression;
+            None for unsupervised learning.
+        """
+        # Get feature_sizes in whole numbers
+        n_features = X.shape[-1]
+        if np.issubdtype(self.feature_sizes.dtype, np.integer):
+            assert (self.feature_sizes >= 0).all(), 'Expected feature sizes in [0, n_features]'
+            assert (self.feature_sizes <= n_features).all(), 'Expected feature sizes in [0, n_features]'
```
Comment (bbengfort) [inline yellowbrick/model_selection/dropping_curve.py:178]: In general, I'd prefer that we did not use assertions to raise `AssertionError` types, we generally expect that if Yellowbrick is raising an exception it is a subclass of [`YellowbrickError`](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/exceptions.py#L21) - here a `YellowbrickTypeError` or `YellowbrickValueError` is probably most appropriate.

File: yellowbrick/model_selection/dropping_curve.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    -----
+    This visualizer is based on sklearn.model_selection.validation_curve
+    """
+
+    def __init__(
+        self,
+        estimator,
+        feature_sizes=DEFAULT_FEATURE_SIZES,
+        groups=None,
+        ax=None,
```
Comment (bbengfort) [inline yellowbrick/model_selection/dropping_curve.py:128]: Other [Yellowbrick model visualizers](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/classifier/rocauc.py#L66-L76) have the `estimator` then the `ax` as the first two arguments, saving visualizer-specific arguments until after these positions (but leaving `**kwargs` last). Although `ax` is not a positional argument, this inconsistency may cause users of this visualizer to make mistakes when shortcutting the call with positional args.

File: yellowbrick/model_selection/dropping_curve.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        """
+        # Get feature_sizes in whole numbers
+        n_features = X.shape[-1]
+        if np.issubdtype(self.feature_sizes.dtype, np.integer):
+            assert (self.feature_sizes >= 0).all(), 'Expected feature sizes in [0, n_features]'
+            assert (self.feature_sizes <= n_features).all(), 'Expected feature sizes in [0, n_features]'
+            self.feature_sizes_ = self.feature_sizes
+        else:
+            assert (self.feature_sizes >= 0.0).all(), 'Expected feature ratio in [0, 1]'
+            assert (self.feature_sizes <= 1.0).all(), 'Expected feature ratio in [0, 1]'
```
Comment (bbengfort) [inline yellowbrick/model_selection/dropping_curve.py:182]: See above.

File: yellowbrick/model_selection/dropping_curve.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            self.feature_sizes_ = np.ceil(n_features * self.feature_sizes).astype(int)
+
+        # The easiest way to prepend a random-dropout layer is to use
+        # SelectKBest with a random scoring function.
```
Comment (bbengfort) [inline yellowbrick/model_selection/dropping_curve.py:186]: Good comment - and cool use of `SelectKBest`

File: tests/test_model_selection/test_dropping_curve.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+except ImportError:
+    pd = None
+
+
+##########################################################################
+# Test Cases
+##########################################################################
+
+
+@pytest.mark.usefixtures("classification", "regression", "clusters")
```
Comment (bbengfort) [inline tests/test_model_selection/test_dropping_curve.py:50]: Is `"clusters"` necessary for testing a dropping curve?

**Review by bbengfort — APPROVED** [review]
This is a well-put-together visualizer! The visualizer pattern makes sense and effectively wraps scikit-learn and matplotlib to produce a dropping curve for both regression and classification models. The tests are well done - they cover most of the visualizer and explore both classifiers, regressors, default options, and extended options in a way that gives me confidence that they protect the visualizer from breaking changes in the future. This is effective library code! 

 I had a couple of minor comments that are easily sorted. Thank you so much for your contribution to Yellowbrick! @lwgray I'll let you take it from here.

**Comment by charlesincharge** [comment]
Awesome, thanks for all the help! Hope this is useful for everyone!

**Comment by lwgray** [comment]
@charlesincharge This is a wonderful visualizer.  Do you think you can do the docs?

**Comment by charlesincharge** [comment]
Thanks! Yes, I'll write some docs!

**Comment by lwgray** [comment]
Can you reference https://github.com/DistrictDataLabs/yellowbrick/issues/1235 in your PR?

**Pull request merged**